server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/QuotaChangesMethod.scala

```diff
@@ -0,0 +1,102 @@
+                                   val quotaManager: QuotaManager,
+                                   val quotaRootResolver: UserQuotaRootResolver) extends MethodRequiringAccountId[QuotaChangesRequest] {
+
+  override val methodName: Invocation.MethodName = MethodName("Quota/changes")
+
+  override val requiredCapabilities: Set[CapabilityIdentifier] = Set(JMAP_QUOTA, JMAP_CORE)
+
+  val quotaChangesResolver: QuotaChangesResolver = QuotaChangesResolver(quotaManager, quotaRootResolver)
+
+  override def doProcess(capabilities: Set[CapabilityIdentifier], invocation: InvocationWithContext, mailboxSession: MailboxSession, request: QuotaChangesRequest): Publisher[InvocationWithContext] =
+    quotaChangesResolver.getLatestState(mailboxSession.getUser, capabilities)
+      .map(newState => QuotaChangesResponse.from(request.sinceState, UuidState.fromJava(newState), request.accountId))
+      .map(response => InvocationWithContext(
+        invocation = Invocation(
+          methodName = methodName,
+          arguments = Arguments(QuotaSerializer.serializeChanges(response)),
+          methodCallId = invocation.invocation.methodCallId),
+        processingContext = invocation.processingContext))
+      .onErrorResume {
+        case e: CanNotCalculateChangesException => SMono.just(InvocationWithContext(Invocation.error(ErrorCode.CannotCalculateChanges, e.getMessage, invocation.invocation.methodCallId), invocation.processingContext))
+        case e => SMono.error(e)
+      }
+
+  override def getRequest(mailboxSession: MailboxSession, invocation: Invocation): Either[Exception, QuotaChangesRequest] =
+    QuotaSerializer.deserializeQuotaChangesRequest(invocation.arguments.value) match {
+      case JsSuccess(mailboxGetRequest, _) => Right(mailboxGetRequest)
+      case errors: JsError => Left(new IllegalArgumentException(ResponseSerializer.serialize(errors).toString))
+    }
+
+}
+
+case class QuotaChangesResolver(private var quotaManager: QuotaManager,
+                                private var quotaRootResolver: UserQuotaRootResolver) {
+
+  private val jmapQuotaManagerWrapper: JmapQuotaManagerWrapper = JmapQuotaManagerWrapper(quotaManager, quotaRootResolver)
+
+  def getLatestState(username: Username, capabilities: Set[CapabilityIdentifier]): SMono[State] =
+    jmapQuotaManagerWrapper.retrieveQuotaRoot(username, capabilities)
+      .flatMap(quotaRoot => SMono(quotaManager.getQuotasReactive(quotaRoot)))
+      .collectSeq()
+      .map(quotasList => evaluateStateFromQuotas(username, capabilities, quotasList))
+
+
+  private def evaluateStateFromQuotas(username: Username, capabilities: Set[CapabilityIdentifier], quotasList: Seq[QuotaManager.Quotas]): State = {
+    if (!quotasList.exists(quotas => quotas.getMessageQuota.getLimit.isLimited || quotas.getStorageQuota.getLimit.isLimited)) {
+      throw new CanNotCalculateChangesException("Quota is empty, hence we cannot calculate changes")
```

Comment:
I am not sure we can use `cannotCalculateChanges` like that.

```
`cannotCalculateChanges`: The server cannot calculate the changes from the state string given by the client. Usually, this is due to the client’s state being too old or the server being unable to produce an update to an intermediate state when there are too many updates. The client MUST invalidate its Foo cache.
```

I think the case a user has an unlimited quota unchanged, we should find a way to return noChanges instead of throwing an error.

Reply:
```
I think the case a user has an unlimited quota unchanged, we should find a way to return noChanges instead of throwing an error.
```
I'm waiting for another opinion!